libafl_qemu/src/modules/usermode/redirect_stdin.rs

```diff
@@ -0,0 +1,140 @@
+    total: usize,
+}
+
+impl RedirectStdinModule {
+    #[must_use]
+    /// constuctor
+    pub fn new() -> Self {
+        Self::with_input_addr(None)
+    }
+
+    #[must_use]
+    /// Create with specified input address
+    pub fn with_input_addr(addr: Option<GuestAddr>) -> Self {
+        Self {
+            input_addr: addr,
+            read: 0,
+            total: 0,
+        }
+    }
+
+    /// Tell this module where to look for the input addr
+    pub fn set_input_addr(&mut self, addr: Option<GuestAddr>) {
+        self.input_addr = addr;
+    }
+}
+
+impl<I, S> EmulatorModule<I, S> for RedirectStdinModule
+where
+    I: Unpin + HasLen + Debug,
+    S: Unpin,
+{
+    fn first_exec<ET>(
+        &mut self,
+        _qemu: Qemu,
+        emulator_modules: &mut EmulatorModules<ET, I, S>,
+        _state: &mut S,
+    ) where
+        ET: EmulatorModuleTuple<I, S>,
+    {
+        emulator_modules.pre_syscalls(Hook::Function(syscall_read_hook::<ET, I, S>));
+    }
+
+    fn pre_exec<ET>(
+        &mut self,
+        _qemu: Qemu,
+        _emulator_modules: &mut EmulatorModules<ET, I, S>,
+        _state: &mut S,
+        input: &I,
+    ) where
+        ET: EmulatorModuleTuple<I, S>,
+    {
+        assert!(self.input_addr.is_some());
+        self.total = input.len();
+        self.read = 0;
+    }
+}
+
+#[expect(clippy::too_many_arguments)]
+fn syscall_read_hook<ET, I, S>(
+    _qemu: Qemu,
+    emulator_modules: &mut EmulatorModules<ET, I, S>,
+    _state: Option<&mut S>,
+    syscall: i32,
+    x0: GuestAddr,
+    x1: GuestAddr,
+    x2: GuestAddr,
+    _x3: GuestAddr,
+    _x4: GuestAddr,
+    _x5: GuestAddr,
+    _x6: GuestAddr,
+    _x7: GuestAddr,
+) -> SyscallHookResult
+where
+    ET: EmulatorModuleTuple<I, S>,
+    I: Unpin + HasLen + Debug,
+    S: Unpin,
+{
+    let h = emulator_modules.get_mut::<RedirectStdinModule>().unwrap();
+    let Some(addr) = h.input_addr else {
+        return SyscallHookResult::new(None);
+    };
+    if syscall == SYS_read as i32 && x0 == 0 {
+        /*
+        println!(
+            "Is sys read {:x} {} {:x} {:x} {} {} {} {} {}",
+            rip, x0, x1, x2, x3, x4, x5, x6, x7
+        );
+        */
+        let size = unsafe {
+            let mut src = addr as *const u8;
+            src = src.wrapping_add(h.read);
+            let dst = x1 as *mut u8;
+            if h.total >= h.read {
+                let size = std::cmp::min(x2, (h.total - h.read).try_into().unwrap());
+                /*
+                println!(
+                    "trying to read {} bytes copying src: {:p} {:p} size: {} h.total: {} h.read {} ",
+                    x2, src, dst, size, h.total, h.read
+                );
+                */
+                dst.copy_from(src, size as usize);
+                size
+            } else {
+                0
+            }
+        };
+        // println!("copied {}", size);
+        h.read += size as usize;
+        return SyscallHookResult::new(Some(size));
```

Comment:
@rmalmain I think this API should be either return `::hooked(val)` or `::skipped`(?)